controllers/argocd/role.go

```diff
@@ -156,6 +181,114 @@ func (r *ReconcileArgoCD) reconcileRole(name string, policyRules []v1.PolicyRule
+				break
+			}
+		}
+		if managedNamespace {
+			continue
+		}
+
+		log.Info(fmt.Sprintf("Reconciling role for %s", namespace.Name))
+
+		role := newRoleForSupportedNamespaces(name, namespace.Name, policyRules, cr)
+		if err := applyReconcilerHook(cr, role, ""); err != nil {
+			return nil, err
+		}
+		role.Namespace = namespace.Name
+		existingRole := v1.Role{}
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: role.Name, Namespace: namespace.Name}, &existingRole)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return nil, fmt.Errorf("failed to reconcile the role for the service account associated with %s : %s", name, err)
+			}
+			log.Info(fmt.Sprintf("creating role %s for Argo CD instance %s in namespace %s", role.Name, cr.Name, namespace))
+			if err := r.Client.Create(context.TODO(), role); err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		// do not reconcile roles for namespaces already containing managed-by-cluster-argocd label
+		// as it already contains roles reconciled during reconcilation of ManagedNamespaces
+		for _, ns := range managedClusterArgoCDNamespaces.Items {
+			if reflect.DeepEqual(ns, namespace) {
+				managedNamespace = true
+				break
+			}
+		}
+		if managedNamespace {
+			continue
+		}
+
+		// reconcile roles only if another ArgoCD instance is not already set as value for managed-by-cluster-argocd label
+		if value, ok := namespace.Labels[common.ArgoCDManagedByClusterArgoCDLabel]; !ok || value != "" {
+			if value != "" && value != cr.Name {
+				log.Info(fmt.Sprintf("Namespace already has label set to argocd instance %s. Thus, skipping namespace %s", value, namespace.Name))
+				continue
+			}
+
+			// Get the latest value of namespace before updating it
+			if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: namespace.Name}, namespace); err != nil {
+				return nil, err
+			}
+			// Update namespace with managed-by-cluster-argocd label
+			namespace.Labels[common.ArgoCDManagedByClusterArgoCDLabel] = cr.Name
+			if err := r.Client.Update(context.TODO(), namespace); err != nil {
+				log.Error(err, fmt.Sprintf("failed to add label from namespace [%s]", namespace.Name))
+			}
+			// if the Rules differ, update the Role
+			if !reflect.DeepEqual(existingRole.Rules, role.Rules) {
+				existingRole.Rules = role.Rules
+				if err := r.Client.Update(context.TODO(), &existingRole); err != nil {
+					return nil, err
+				}
+			}
+			roles = append(roles, &existingRole)
+
+			value, ok := r.ManagedSourceNamespaces[cr.Name]
+			if !ok {
+				value = &corev1.NamespaceList{
+					Items: []corev1.Namespace{*namespace},
+				}
+			} else {
+				alreadyManagedSourceNS := false
+				for _, n := range value.Items {
+					if n.Name == namespace.Name {
+						alreadyManagedSourceNS = true
+						break
+					}
+				}
+				if !alreadyManagedSourceNS {
+					value.Items = append(value.Items, *namespace)
+				}
+			}
+			r.ManagedSourceNamespaces[cr.Name] = value
```

Comment:
If we were to follow what I suggested in the earlier comment, this part of the code would not be needed 
`r.ManagedSourceNamespaces` could be populated the same way as is done [here](https://github.com/argoproj-labs/argocd-operator/blob/90eadf8356c81bf66bd1928d2354fc6e34aed3ef/controllers/argocd/util.go#L1347) for `r.ManagedNamespaces` i.e, populate the map/list each time there is a reconciliation done for a specific argocd instance in question